File: Modelo/Simulacion.java
package Modelo;
import java.sql.SQLException;
import java.text.DecimalFormat;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Scanner;

import DAORelacional.DAOSimulacion;

public class Simulacion {

	private Integer id=0;					// PK en base de datos
	private LocalDateTime timestamp;		// Fecha de la simulación
	private SimEngine se;					// Motor de simulación que se utiliza
	private Lander lander;					// Módulo que se utiliza
	private Player user;					// Jugador que la efectúa
	private Escenario planet;				//	Escenario de la simulación
	private ArrayList<DatosSim> simData = new ArrayList<DatosSim>();	// Datos de la simulación
	private static boolean outOfFuel = false;
	public boolean __break = false;			// Flag para terminar la simulación ( el usuario abandona)
	// Para abandonar hay que introducir un nivel de impulso = -1
	
	// Constructor
	public Simulacion(Player _user, Lander lander, Escenario planet) {
		super();
		this.lander = lander;
		this.user = _user;
		this.planet = planet;
		this.timestamp = LocalDateTime.now();
		init();
	}
	// Getter y Setter
	public LocalDateTime getTimestamp() 				{		return timestamp;	}
	public void setTimestamp(LocalDateTime timestamp) 	{		this.timestamp = timestamp;	}
	public SimEngine getSe() 							{		return se;	}
	public void setSe(SimEngine se) 					{		this.se = se;	}
	public Lander getLander() 							{		return lander;	}
	public void setLander(Lander lander) 				{		this.lander = lander;	}
	public Player getUser() 							{		return user;	}
	public void setUser(Player user) 					{		this.user = user;	}
	public Escenario getPlanet() 						{		return planet;	}
	public void setPlanet(Escenario planet) 			{		this.planet = planet;	}
	public ArrayList<DatosSim> getSimData()	 			{		return simData;	}
	public void setSimData(ArrayList<DatosSim> simData) {		this.simData = simData;	}
	public Integer getId() 								{		return id;	}
	public void setId(Integer id) 						{		this.id = id;	}
	
	// Métodos
	
	// Añade un instante de simulación a la simulación
	// Los datos vendrán de SimEngine, sin el dato de fuel en depósito.
	public void addSimData() {
		DatosSim ds = se.getSimData();
		ds.setFuel(getLander().getFuel_deposito());
		simData.add(ds);
	}
	
	/**
	 * Inicializa la simulación
	 * @return
	 */
	public void init() {
		// Valores iniciales
		this.se = new SimEngine(planet.getHe(),planet.getVe(),planet.getG());
	}
	
	/**
	 * Muestra panel de indicadores
	 */
	public void muestraPanel() {
		Double vel = se.getVel();
		Double dist =se.getDist();
		Integer tiempo = se.getTiempo();
		Double fuel_deposito = lander.getFuel_deposito();
		DecimalFormat df = new DecimalFormat("+0000.00;-0000.00");
		
		if (tiempo == 0) { // inicio de la simulación 
	        System.out.println("TIEMPO  DISTANCIA   VEL        FUEL      NIVEL IMPULSO"); // Tablero de indicadores
	        System.out.println("------------------------------------------------------");
		}
		 // Muestra valores de los parámetros de la simulación, formateados en columnas
         System.out.printf("%03d    %s    %s    %04d        ",
        		           tiempo,df.format(dist),df.format(vel),fuel_deposito.intValue()
        		          );
	}
	
	public void aplicaMotor(Lander l) {
		
		Double impulso=0.0;
		Integer nivel_impulso = 0;
		
		// Sólo si tenemos fuel para quemar
		if (!outOfFuel) {
			Scanner sc = new Scanner(System.in);
	        System.out.print("¿(0-9)? >");                      	// Solicita nivel de impulso       
			nivel_impulso = sc.nextInt();                       	// Lectura de teclado
			if (nivel_impulso ==-1) {
				nivel_impulso =0;
				__break = true;										// Abandonar la simulación
			}
			if (nivel_impulso <0) nivel_impulso =0;					// Sencilla comprobación de límites
			if (nivel_impulso >9) nivel_impulso =9;
			
		}
		else {
			System.out.println("SIN FUEL , CAIDA LIBRE!");
			try {
				Thread.sleep(1000);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		
		if (lander.getFuel_deposito() == 0) nivel_impulso =0;   // Si no queda fuel , no tiene efecto la elección
		impulso = lander.getPerfPot(nivel_impulso);        		// Elijo, en función del nivel el impulso instantáneo
		se.setImpulso(impulso);									// Pasar el impulso al motor de simulación.
		// Consumo de combustible
        lander.setFuel_a_quemar( impulso * 2);      			// No es una simulación realista
        lander.setFuel_deposito(lander.getFuel_deposito() - lander.getFuel_a_quemar());      // Actualizo la reserva de fuel
		if (lander.getFuel_deposito()<0) 	{					// Eliminar incosistencias en el cálculo
				lander.setFuel_deposito(0);  
				outOfFuel = true;
		} // Sin fuel
	}
	
	public Boolean show_result() {
		
		Boolean flag = false;				// Terminación sin salvar puntuación
        // Comprobar la condiciones de aterrizaje y mostrar información sobre el mismo.
		Double vel_fin = se.getVel();
		Double dist_fin =se.getDist();
		Integer tiempo = se.getTiempo();
		Double fuel_deposito = lander.getFuel_deposito();
		DecimalFormat df = new DecimalFormat("+0000.00;-0000.00");
		
		if (!this.__break) {
		
			flag = (Math.abs(se.getVel())>lander.getRes_tren());
			if (flag){
                System.out.println("\nHAS ESTRELLADO LA NAVE");
                System.out.println("------------------------------------------------");
                System.out.println("VELOCIDAD DE ENTRADA    : "+ df.format(vel_fin) + " m/s");
                System.out.println("HAS HECHO UN CRATER DE  : "+ df.format(Math.abs(dist_fin)) + " m");
                System.out.println("------------------------------------------------");
                flag = false;	// no salvar
			}
			else {
                System.out.println("\nATERRIZAJE EXITOSO!!");
                System.out.println("------------------------------------------------");
                System.out.println("TIEMPO DE ATERRIZAJE : " + tiempo + " s");
                System.out.println("FUEL EN DEPOSITO     : " + fuel_deposito + " l");
                System.out.println("------------------------------------------------");
                flag=true;		// salvar
			}
		}
		else {
		    System.out.println("\nSIMULACIÓN INTERRUMPIDA POR EL USARIO");
            System.out.println("------------------------------------------------");
            System.out.println("TIEMPO DE SIMULACIÓN : " + tiempo + " s");
            System.out.println("FUEL EN DEPOSITO     : " + fuel_deposito + " l");
            System.out.println("DISTANCIA A OBJETIVO : " + dist_fin + " m");
            System.out.println("------------------------------------------------");	
            flag = false;	// no salvar
		} // Misión finalizada o interrumpida
	 return flag;	
	}  // show_result
	
	/**
	 * Salva los datos de simulación en base de datos
	 * @return
	 */
	public boolean saveSim(String Modo) {
		DAOSimulacion ds = new DAOSimulacion(Modo);
		try {
			if (ds.saveSimulacion(this))
				System.out.println("Datos Almacenados en base de datos.");
			ds._c.close();
		} catch (SQLException e) {
			e.printStackTrace();
		}
		return true;
	}
	
	/**
	 * Registra el marcador obtenido en la base de datos
	 * Es llamado de forma interna por saveSim()
	 * @return
	 */
	private boolean saveScore() {
		return true;
	}
	
	public void show() {
		
		// Salida por pantalla;
	}
}

File: Modelo/Simulacion.php
<?php
class Simulacion{
    private $id=0; //PK en base de datos
    private $timestamp; // Feccha de simulación
    private $se; // Motor de simulaciónque se utiliza
    private $lander; //Modulo que se utiliza
    private $user; //Jugador que la afecta
    private $planet; // Escenario de la simulación
    private $simData = array(); //Datos de la simulación
    private static $outOfFuel=false; 
    private $__break = false; // Flag para terminar la simulación (el usuario abandona)
    // Para abandonar hay que introducir un nivel de impulso = -1

    // Constructor
    public function __Simulacion($_user,$lander,$planet){
        $this->user=$_user;
        $this->lander=$lander;
        $this->planet=$planet;
        $this->timestamp=new DateTime();
    }
    public function getTimeStamp(){
        return $this->timestamp;
    }
    public function setTimeStamp($timestamp){
        $this->timestamp=$timestamp;
    }
    public function getSe(){
        return $this->se;
    }
    public function setSe($se){
        $this->se=$se;
    }
    public function getLander(){
        return $this->lander;
    }
    public function setLander($lander){
        $this->lander=$lander;
    }
    public function getUser(){
        return $this->user;
    }
    public function setUser($user){
        $this->user=$user;
    }
    public function getPlanet(){
        return $this->planet;
    }
    public function setPlanet($planet){
        $this->planet=$planet;
    }
    public function setSimData(array $simData){
        $this->simData=$simData;
    }
    public function getId(){
        return $this->id;
    }
    public function setId($id){
        $this->id=$id;
    }
    public function addSimData(){
        $ds = $this->se->getSimData();
        $ds->setFuel($this->getLander()->getFuel());
        $this->simData[]=$ds;
    }
    public function init():void{
        $this->se = new SimEngine($this->planet->getHe(), $this->planet->getVe(), $this->planet->get_G());
    }
    public function muestraPanel(){
        $vel = $this->se->getVel();
        $dist = $this->se->get_dist();
        $tiempo = $this->se->getTiempo();
        $fuel_deposito = $this->lander->get_fuel_deposito();

        $numbeR = 123.456;
        $formattNumber = sprintf("%+07.2f", $numbeR);
        echo $formattNumber."\n";

        $number = -123.456;
        $formattNumber = sprintf("%+07.2f", $number);
        echo $formattNumber."\n";

        if ($tiempo==0) {
            echo "TIEMPO  DISTANCIA   VEL        FUEL      NIVEL IMPULSO"."\n";
            echo "-------------------------------------------------------"."\n";
        }

        $formattedDist = sprintf("%+07.2f", $dist);
        $formattedVel = sprintf("%+07.2f", $vel);

        printf("%03d %s %s %04d", $tiempo, $formattedDist, $formattedVel, intval($fuel_deposito));
    }
    public function aplicaMotor($l){

        $impulso = 0.0;
        $nivel_impuldo = 0;

        // Sólo si tenemos fuel para quemar
        if (!Simulacion::$outOfFuel) {
            $input = trim(fgets(STDIN));
            printf("?(0-9)"); //Solicita nivel de impulso
            $nivel_impulso = $input; // Lectura de teclado
            if ($nivel_impulso == -1) {
                $nivel_impulso = 0;
                $__break = true;      // Abandonar la simulación
            }
            if ($nivel_impulso < 0) $nivel_impulso = 0; // Sencilla comprobación de límites
            
            if ($nivel_impulso > 0) $nivel_impulso = 9;
        }
        else{
            printf("SIN FUEL , CAIDA LIBRE!");
            sleep(1);
        }

        if ($this->lander->getFuel_deposito() == 0) $nivel_impulso = 0; // Si no queda fuel, no  tiene efecto la eleccón
        $impulso = $this->lander->getPerfPot($nivel_impulso);            // Elijo, en función del nivel el impulso instantaneo
        $this->se->set_impulso($impulso);                                // Pasar el impulso al motot de simulación
        //Consumo de combustible
        $this->lander->setFuel_a_quemar($impulso * 2);                   // No es una simulación realista
        $this->lander->setFuel_deposito($this->lander->getFuel_deposito() - $this->lander->getFuel_a_quemar()); // Actualizo la reserva de fuel

        if ($this->lander->getFuel_deposito() < 0){                      // Eliminar inconsistencias en el cálculo
            $this->lander->setFuel_deposito(0);
            $outOfFuel = true;
        } // Sin fuel
    }

    public function show_result(){
        $flag = false;         // Terminación sin salvar puntación
        $vel_fin = $this->se->getVel();
        $dist_fin = $this->se->get_dist();
        $tiempo = $this->se->get_tiempo();
        $fuel_deposito = $this->lander->getFuel_deposito();
        
        $formattedVel_fin = sprintf("%+07.2f", $vel_fin);
        $formattedDist_fin = sprintf("%+07.2f", $dist_fin);

        echo "Tiempo: $tiempo\n";
        echo "Velocidad final: $formattedVel_fin\n";
        echo "Distancia final: $formattedDist_fin\n";
        echo "Fuel en depósito: " . intval($fuel_deposito) . "\n";
        
        if ($flag === false) {
            echo "Terminación sin salvar la puntuación";
        }

        if (!$this->__break) {
            $flag = (abs($this->se->getVel()>$this->lander->getRes_tren()));
            if ($flag) {
                echo "HAS ESTRELLADO LA NAVE\n";
                echo "------------------------------------------------\n";
                echo "VELOCIDAD DE ENTRADA    : " . sprintf("%+07.2", $vel_fin) . "m/s\n";
                echo "HAS HECHO UN CRATER DE  : ". sprintf("%+07.2", abs($dist_fin)) . " m\n";
                echo "------------------------------------------------\n";
                $flag = false; // no slavar
            }
            else{
                echo "\nATERRIZAJE EXITOSO!!\n";
                echo "------------------------------------------------\n";
                echo "TIEMPO DE SIMULACIÓN : " . $tiempo . " s\n";
                echo "FUEL EN DEPOSITO     : " . $fuel_deposito . " l\n";
                echo "DISTANCIA A OBJETIVO : " . $dist_fin . " m\n";
                echo "------------------------------------------------";
                $flag = false; // no salvar
            } // Misión finalizaada o interrumpida
            return $flag;
        } // show_result

        /**
         * Salvar los datos de simulación en base de datos
         * @return
         */
    }
    public function saveSim($Modo){
            $ds = new DAOSimulacion($Modo);

            try{
                if ($ds->saveSimulacion($this)) {
                    echo "Datos Almacenados en base de datos.\n";
                }
                $ds->_c=null;
            } catch (PDOException $e) {
                echo "Error: " . $e->getMessage();
            }
            return true;
    }
    /**
     * Registra el marcador obtenido en la base de datos
     * Es llamado de forma interna por saveSim()
     * @return 
     */

     private function saveScore(){
        return true;
     }
     public function show(){
        //Salida por pantalla
     }
}
?>